pkg/sql/parser/sql.y

```diff
@@ -3797,6 +3837,251 @@ create_extension_stmt:
+    $$.val = arg
+  }
+
+func_arg:
+  func_arg_class param_name func_arg_type
+  {
+    $$.val = tree.FuncArg{
+      Name: tree.Name($2),
+      Type: $3.typeReference(),
+      Class: $1.functionArgClass(),
+    }
+  }
+| param_name func_arg_class func_arg_type
+  {
+    $$.val = tree.FuncArg{
+      Name: tree.Name($1),
+      Type: $3.typeReference(),
+      Class: $2.functionArgClass(),
+    }
+  }
+| param_name func_arg_type
+  {
+    $$.val = tree.FuncArg{
+      Name: tree.Name($1),
+      Type: $2.typeReference(),
+      Class: tree.FunctionArgIn,
+    }
+  }
+| func_arg_class func_arg_type
+  {
+    $$.val = tree.FuncArg{
+      Type: $2.typeReference(),
+      Class: $1.functionArgClass(),
+    }
+  }
+| func_arg_type
+  {
+    $$.val = tree.FuncArg{
+      Type: $1.typeReference(),
+      Class: tree.FunctionArgIn,
+    }
+  }
+
+func_arg_class:
+  IN { $$.val = tree.FunctionArgIn }
+| OUT { return unimplemented(sqllex, "create function with 'OUT' argument class") }
+| INOUT { return unimplemented(sqllex, "create function with 'INOUT' argument class") }
+| IN OUT { return unimplemented(sqllex, "create function with 'IN OUT' argument class") }
+| VARIADIC { return unimplemented(sqllex, "create function with 'VARIADIC' argument class") }
+
+func_arg_type:
+  typename
+
+func_return_type:
+  func_arg_type
+
+opt_create_func_opt_list:
+  create_func_opt_list { $$.val = $1.functionOptions() }
+| /* EMPTY */ { $$.val = tree.FunctionOptions{} }
+
+create_func_opt_list:
+  create_func_opt_item { $$.val = tree.FunctionOptions{$1.functionOption()} }
+| create_func_opt_list create_func_opt_item { $$.val = append($1.functionOptions(), $2.functionOption())}
+
+create_func_opt_item:
+  AS func_as
+  {
+    $$.val = tree.FunctionBodyStr($2)
+  }
+| LANGUAGE non_reserved_word_or_sconst
+  {
+    lang, err := tree.AsFunctionLanguage($2)
+    if err != nil {
+      return setErr(sqllex, err)
+    }
+    $$.val = lang
+  }
+| TRANSFORM { return unimplemented(sqllex, "create function...transform") }
+| WINDOW { return unimplemented(sqllex, "create function...window") }
+| common_func_opt_item
+  {
+    $$.val = $1.functionOption()
+  }
+
+common_func_opt_item:
+  CALLED ON NULL INPUT
+  {
+    $$.val = tree.FunctionCalledOnNullInput
+  }
+| RETURNS NULL ON NULL INPUT
+  {
+    $$.val = tree.FunctionReturnsNullOnNullInput
+  }
+| STRICT
+  {
+    $$.val = tree.FunctionStrict
+  }
+| IMMUTABLE
+  {
+    $$.val = tree.FunctionImmutable
+  }
+| STABLE
+  {
+    $$.val = tree.FunctionStable
+  }
+| VOLATILE
+  {
+    $$.val = tree.FunctionVolatile
+  }
+| EXTERNAL SECURITY DEFINER
+  {
+    return unimplemented(sqllex, "create function...security")
+  }
+| EXTERNAL SECURITY INVOKER
+  {
+    return unimplemented(sqllex, "create function...security")
+  }
+| SECURITY DEFINER
+  {
+    return unimplemented(sqllex, "create function...security")
+  }
+| SECURITY INVOKER
+  {
+    return unimplemented(sqllex, "create function...security")
+  }
```

Comment:
One downside I could see of returning these errors at parse time is I'm not sure we'll get telemetry for the feature we don't support. Maybe create a follow-up issue to make sure we get good telemetry out of unsupported features?

Reply:
ah, good point, I created #83964 to create this